### DISCUSSION TITLE
Determine when uninstall stanza is run explicitly by user?

### DISCUSSION THREAD

**Discussion by mschrage** [Homebrew/discussions#3401] [Writing Formulae/Casks] [answered]
I'm the maintainer of the [`fig`](https://github.com/Homebrew/homebrew-cask/blob/HEAD/Casks/fig.rb) Cask and I've been trying to find a way to replicate the behavior of our app uninstall flow. 

### Context
If a user runs `fig uninstall` or moves the app to the trash, we take the user to [fig.io/uninstall](fig.io/uninstall) in order to gather some additional feedback.

I've tried to replicate this behavior in our cask as part of the `fig app uninstall` script. In the process, I've realized that the `uninstall`  stanza is invoked in many situations, not just when the user runs `brew uninstall fig`.

### Question

Is there any way to distinguish when uninstall stanza is run explicitly by user as in `brew uninstall fig` vs when it is run _on behalf_ of the user, for instance, when the user runs `brew upgrade` or brew `brew reinstall fig`.

If not, is this functionality you'd be open to supporting? (I could imagine this information exposed as an environment variable that is set while the `script` and `early_script` are executed).

**Comment by SMillerDev** [comment]
No, that's what zap stanzas are for.

**Reply by mschrage** [reply]
I don't think it is possible to do what we want with a zap stanza, since we can't handle the case where the user runs `brew uninstall fig` any differently from `brew upgrade`? 

My sense is that many (most?) users uninstall casks by running `brew uninstall <cask>` and omit the `--zap`.

**Comment by carlocab** [comment, marked as answer]
Homebrew Cask was designed to allow you to manage your installations purely from a CLI, so taking users away from the CLI when they don't expect probably isn't welcome, and might cause errors in certain situations (e.g. when logged in via `ssh` with no attached GUI). So I don't think supporting this is something we can do.

I'm not opposed to adding a message in the `caveats`, though. But this does have the shortcoming of not being displayed after `brew uninstall`.
